syntax = "proto3";
package auth;
option go_package = "com.github/qcodelabsllc/qiggy;ms";
option java_multiple_files = true;
option java_package = "io.qcodelabsllc.qiggy";
option java_outer_classname = "QiggyAuthService";

import "google/protobuf/wrappers.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

service AuthService {
  rpc login(LoginRequest) returns (AuthResponse);
  rpc create_account(CreateAccountRequest) returns (AuthResponse);
  rpc reset_password(google.protobuf.StringValue) returns (google.protobuf.StringValue);
  rpc send_auth_code(google.protobuf.StringValue) returns (AuthCodeResponse);
  rpc verify_auth_code(VerifyAuthCode) returns (AuthCodeResponse);
  rpc update_password(google.protobuf.StringValue) returns (google.protobuf.StringValue);
  rpc update_account(Account) returns (Account);
  rpc logout(google.protobuf.StringValue) returns (google.protobuf.Empty);
  rpc refresh_token(google.protobuf.StringValue) returns (Session);
  rpc verify_token(google.protobuf.StringValue) returns (google.protobuf.BoolValue);

  // experimental rpc
  rpc verify_phone_number(stream AuthCodeRequest) returns (stream AuthCodeResponse);
}

enum PhoneVerificationStatus {
  none = 0;
  code_sent = 1;
  verified = 2;
  error = 3;
}

message VerifyAuthCode {
  int32 code = 1;
  string phone_number = 2;
}

message AuthCodeRequest {
  int32 code = 1;
  string phone_number = 2;
}

message AuthCodeResponse {
  PhoneVerificationStatus status = 1;
  bool successful = 2;
  string message = 3;
}

message AuthResponse {
  Account account = 1;
  Session session = 2;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message CreateAccountRequest {
  string username = 1;
  string password = 2;
  string display_name = 3;
  string phone_number = 4;
}

message Account {
  string username = 1;
  string password = 2;
  string display_name = 3;
  string phone_number = 4;
}

message Session {
  string access_token = 1;
  string refresh_token = 2;
  google.protobuf.Timestamp access_token_expiry = 3;
  google.protobuf.Timestamp refresh_token_expiry = 4;
  string session_id = 5;
  google.protobuf.Timestamp created_at = 6 [lazy = true];
}